libs/basics/src/iterators/lines.ts

```diff
@@ -0,0 +1,136 @@
+    (function* gen() {
+      let hasSeenAnyChunks = false;
+
+      for (const element of iterable) {
+        const chunk = element.toString();
+        let offset = 0;
+        for (const newlineMatch of matchAllNewlines(chunk)) {
+          /* istanbul ignore next */
+          if (typeof newlineMatch.index !== 'number') {
+            throw new Error('expected index');
+          }
+
+          const line = buffer + chunk.slice(offset, newlineMatch.index);
+          buffer = '';
+          offset = newlineMatch.index + newlineMatch[0].length;
+          yield line;
+        }
+
+        buffer += chunk.slice(offset);
+        hasSeenAnyChunks = true;
+      }
+
+      if (hasSeenAnyChunks) {
+        yield buffer;
+      }
+    })()
+  );
+}
+
+function linesAsync(
+  iterable: AsyncIterable<ToString>
+): AsyncIteratorPlus<string> {
+  let buffer = '';
+  return new AsyncIteratorPlusImpl(
+    (async function* gen() {
+      let hasSeenAnyChunks = false;
+
+      for await (const element of iterable) {
+        const chunk = element.toString();
+        let offset = 0;
+        for (const newlineMatch of matchAllNewlines(chunk)) {
+          /* istanbul ignore next */
+          if (typeof newlineMatch.index !== 'number') {
+            throw new Error('expected index');
+          }
+
+          const line = buffer + chunk.slice(offset, newlineMatch.index);
+          buffer = '';
+          offset = newlineMatch.index + newlineMatch[0].length;
+          yield line;
+        }
+
+        buffer += chunk.slice(offset);
+        hasSeenAnyChunks = true;
+      }
+
+      if (hasSeenAnyChunks) {
+        yield buffer;
+      }
+    })()
+  );
+}
+
+/**
+ * Pipes elements from `iterable` by collecting them together into a string
+ * until a full line is formed, then yielding that line. The final line may or
+ * may not include a newline.
+ *
+ * @example
+ *
+ * ```ts
+ * const file = fs.createReadStream('file.txt', { encoding: 'utf8' });
+ *
+ * expect(await iter(file).lines().toArray()).toEqual([
+ *   'line 1\n',
+ *   'line 2\n',
+ *   …
+ * ]);
+ * ```
+ */
+export function lines(
+  iterable: AsyncIterable<ToString>
+): AsyncIteratorPlus<string>;
+
+/**
+ * Pipes elements from `iterable` by collecting them together into a string
+ * until a full line is formed, then yielding that line. The final line may or
+ * may not include a newline.
+ *
+ * @example
+ *
+ * ```ts
+ * const input = ['hello', ' world', '!', '\n', 'goodbye', 'world'];
+ *
+ * expect(lines(input).toArray()).toEqual([
+ *   'hello world!\n',
+ *   'goodbye world'
+ * ]);
+ * ```
+ */
+export function lines(
+  iterable: ToString | Iterable<ToString>
+): IteratorPlus<string>;
+
+/**
+ * Pipes elements from `iterable` by collecting them together into a string
+ * until a full line is formed, then yielding that line. The final line may or
+ * may not include a newline.
+ */
+export function lines(
+  iterable: ToString | Iterable<ToString> | AsyncIterable<ToString>
```

Comment:
@jonahkagan this version is slightly different from the version that was a method on `IteratorPlus`. This one accepts anything that has a `toString(): string` method as well as iterables of the same. That means it can take `ReadStream`s directly, making `lines(fs.createReadStream(path))` work without any casting.

Reply:
Cool! Maybe worth a test case to demonstrate that use case